hub/test/src/testServer.js

```diff
@@ -10,373 +10,545 @@ import { DriverModel } from '../../src/server/driverModel'
+      t.equal(server.authTimestampCache.cache.itemCount, 4, 'Auth cache should have correct item count')
+      server.authTimestampCache.setupCacheEvictionLogger(1)
+      await new Promise((res) => setTimeout(() => res(), 10))
+      t.equal(server.authTimestampCache.currentCacheEvictions, 0, 'Auth cache eviction count should have been cleared by logger')
+    })
   })
 
-  test('handle scoped writes with association tokens', (t) => {
+  test('fail writes with revoked auth token', async (t) => {
+    await usingMemoryDriver(async (mockDriver) => {
+      const server = new HubServer(mockDriver, new MockProofs(), {
+                                    serverName: TEST_SERVER_NAME,
+                                    authTimestampCacheSize: TEST_AUTH_CACHE_SIZE})
+      const challengeText = auth.getChallengeText(TEST_SERVER_NAME)
+      let authPart = auth.V1Authentication.makeAuthPart(testPairs[0], challengeText)
+      let authorization = `bearer ${authPart}`
+
+      const getJunkData = () => {
+        const s = new Readable()
+        s.push('hello world')
+        s.push(null)
+        return s
+      }
+
+      // no revocation timestamp has been set, write request should succeed
+      await server.handleRequest(testAddrs[0], '/foo/bar', 
+                                { 'content-type' : 'text/text',
+                                  'content-length': 400,
+                                  authorization }, getJunkData())
+
+      // revoke the auth token (setting oldest valid date into the future)
+      const futureDate = (Date.now()/1000|0) + 10000
+      await server.handleAuthBump(testAddrs[0], futureDate, { authorization })
+
+      // write should fail with auth token creation date older than the revocation date 
+      await t.rejects(server.handleRequest(testAddrs[0], '/foo/bar',
+                          { 'content-type' : 'text/text',
+                            'content-length': 400,
+                            authorization }, getJunkData()), errors.AuthTokenTimestampValidationError, 'write with revoked auth token should fail')
+
+      // create a auth token with iat forced further into the future
+      authPart = auth.V1Authentication.makeAuthPart(testPairs[0], challengeText, undefined, undefined, undefined, futureDate + 10000)
+      authorization = `bearer ${authPart}`
+    
+      // request should succeed with a token iat newer than the revocation date
+      await server.handleRequest(testAddrs[0], '/foo/bar', 
+                                { 'content-type' : 'text/text',
+                                  'content-length': 400,
+                                  authorization }, getJunkData())
+    })
+  })
 
-    const writeScopes = [
-      {
-        scope: 'putFile',
-        domain: '/foo/bar',
-      },
-      {
-        scope: 'putFilePrefix',
-        domain: 'baz'
+  test('association tokens with a whitelisted signer can bump any address', async (t) => {
```

Comment:
I don't understand this test -- it seems like the following is happening:

1. An association token is created and included in an authorization token -- this conveys roughly the following:
    Token grants access to write to testPairs[0], testPairs[0] is not in the whitelist, but testPairs[1] is, and grants testPairs[0] the ability to use the gaia hub.
2. That authorization token is used to bump the revocation number for testPairs[0]

Right? In that case, it doesn't really match 'association tokens with a whitelisted signer can bump any address', which isn't behavior we want anyways, right?

Reply:
> Right? In that case, it doesn't really match 'association tokens with a whitelisted signer can bump any address', which isn't behavior we want anyways, right?

No, I think it is.  You need the ability to invalidate auth tokens on a per-app basis.  This can only be achieved on a private Gaia hub (i.e. with a non-empty whitelist) if one of two things are true:

* The app address is whitelisted
* The owner address is whitelisted, and the auth token has an association token authorizing the bump on the app address

This requirement translates to "an auth token with an association token signed by a whitelisted address can be used to bump the auth token's signer's address" -- which is to say, a whitelisted address can bump _any_ address's auth revocation timestamp on the Gaia hub.

Reply:
However, this is _not_ to say that an auth token can be used to bump any address -- regardless of whether or not it has an association token, an auth token can only bump the address of the key that signed it.

Reply:
> This requirement translates to "an auth token with an association token signed by a whitelisted address can be used to bump the auth token's signer's address" -- which is to say, a whitelisted address can bump any address's auth revocation timestamp on the Gaia hub.

Hmm, it still should only ever be the case that a signer can only bump the revocation timestamp for its corresponding address -- and it appears this is what is implemented in the PR. The call to `validate` will only pass if the auth JWT is signed by the private key corresponding to the address being validated.

It's true that if the whitelisted address is a parent key of another address, it'd be able to generate such a token, but that's outside the purview of the authentication scheme in Gaia. But this isn't the case for all addresses, anyways --- imagine there are two whitelisted "owners" on the same private hub -- one owner wouldn't be able to generate auth tokens to bump revocation numbers for app-keys of the other owner.

> However, this is not to say that an auth token can be used to bump any address -- regardless of whether or not it has an association token, an auth token can only bump the address of the key that signed it.

Right, I think that's the same as what I'm saying above.

Reply:
Looking at the test, it's only confirming that a request to bump the non-whitelisted address will be honored as long as the auth token has a well-formed association token signed by a whitelisted address.

Reply:
Yep, I think naming it  "Non-whitelisted address can bump revocation if bearing valid association token" would match what its testing

Reply:
Test renamed